[dtensor] skip pytree when not necessary

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110132

pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

wanchaol committed:
[dtensor] skip pytree when not necessary

pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] skip pytree when not necessary"


pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] skip pytree when not necessary"


pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] skip pytree when not necessary"


pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -82,24 +82,29 @@ def redistribute_local_args(
-            new_flat_local_args.append(reshard_arg_spec)
+            new_local_args.append(reshard_arg_spec)
 
-    op_info.flat_local_args = new_flat_local_args
```
it's interesting it worked without `tuple()` previously..

wanchaol replied:
Yeah it's because we always to unflatten before, which recovers the list to tuple

fduwjj reviewed (approved):
LGTM, it's actually surprising to me that only two ops need treemap here..

wanchaol committed:
Update on "[dtensor] skip pytree when not necessary"


pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

Chillee commented:
Have you tried using optree?

wanchaol committed:
Update on "[dtensor] skip pytree when not necessary"


pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

wanchaol commented:
> Have you tried using optree?

Oh just noticed that got landed!! I can try turn that on, is there a flag to turn it on? 

My feeling is that even if that landed it would still have some overhead exposed (especially optree perf is most amazing on dicts, while operator arguments does not have too many dicts, maybe kwargs is a good try). This PR try to turn pytree/optree off completely for most ops which I thought it would be faster for either case?

Chillee commented:
@wanchaol might be a little bit annoying to actually land but the idea was that we would incrementally port usages over incrementally.

wanchaol commented:
> @wanchaol might be a little bit annoying to actually land but the idea was that we would incrementally port usages over incrementally.

@Chillee sounds good, I'll give it a try in a follow up PR? the annoying part is that we need to guard the imports mostly so that all CI passes?

Chillee commented:
Yeah, see https://github.com/pytorch/pytorch/pull/109684

wanchaol committed:
Update on "[dtensor] skip pytree when not necessary"


pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
